MonoGame.Framework/Graphics/Vertices/IndexBuffer.cs

```diff
@@ -115,10 +115,26 @@ internal protected override void GraphicsDeviceResetting()
         protected void SetDataInternal<T>(int offsetInBytes, T[] data, int startIndex, int elementCount, SetDataOptions options) where T : struct
         {
             if (data == null)
-                throw new ArgumentNullException("data is null");
+                throw new ArgumentNullException("data");
             if (data.Length < (startIndex + elementCount))
                 throw new InvalidOperationException("The array specified in the data parameter is not the correct size for the amount of data requested.");
 
+            var type = typeof (T);
+            switch (IndexElementSize)
+            {
+                case IndexElementSize.SixteenBits:
+                    if (type != typeof(short) || type != typeof(ushort))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when short or ushort need to be used.");
+                    }
+                    break;
+                case IndexElementSize.ThirtyTwoBits:
+                    if (type != typeof(int) || type != typeof(uint))
+                    {
+                        throw new ArrayTypeMismatchException("Invalid type using when int or uint need to be used.");
+                    }
+                    break;
+            }

```

Comment:
> Maybe I add byte and ubyte to exception list ? 

No...  that isn't the fix.

The type of `T` does not matter.  XNA is simply copying an array of bytes as far as it cares.